### ISSUE TITLE
Resource should be destroyed when dropped in the Web backend

### ISSUE THREAD

**Issue by mikialex** [gfx-rs/wgpu#4092]
When testing our rendering engine using wgpu on the web (webgpu) target, we found something that could and should be improved about resource(buffers, textures) management.

In every frame, we create some new texture as render targets in our post-effect system, use them, and drop them immediately. This works fine on the native build, but on the web, we observed a serious memory leak like issue. 

On the Web, the actual GPU resource deallocation happens when the JS web resource objects get garbage collected. So if large-size resources like frame size texture create every frame, due to the  GC trigger time's uncertainty, the peak GPU memory usage will be very high and cause oom problem. 

To workaround this issue, we could call destroy explicitly on the resource, we could reuse the texture every frame. However, to fix this issue directly, I think resources should be destroyed when dropped in the Web backend because it's suited for all cases. Using wasm or rust, means we have explicit control over when our resource should be destroyed(the answer is when dropped) and not rely on the garbage collector at all.

For reference: webgpu spec said in https://www.w3.org/TR/webgpu/#webgpu-interfaces: 

> Calling the destroy method on a [WebGPU interface](https://www.w3.org/TR/webgpu/#webgpu-interface) (such as [GPUDevice](https://www.w3.org/TR/webgpu/#gpudevice).[destroy()](https://www.w3.org/TR/webgpu/#dom-gpudevice-destroy) or [GPUBuffer](https://www.w3.org/TR/webgpu/#gpubuffer).[destroy()](https://www.w3.org/TR/webgpu/#dom-gpubuffer-destroy)) should be favored over relying on garbage collection if predictable release of allocated resources is needed.

**Comment by mikialex** [comment]
After further investigation, I think it's not an available fix. According to the spec, the resource can not be called destroyed if it has been encoded in a command buffer and the command buffer has not been submitted to the queue, so we just can't blindly call destroy when the resource itself dropped. Except we implement another layer of reference counting mechanism in the wgpu layer, which is expensive and not a good solution.

**Comment by ErichDonGubler** [comment]
@mikialex: Would the refcounting being added by #3626 not be sufficient? CC @jimblandy, @gents83.

**Comment by gents83** [comment]
> @mikialex: Would the refcounting being added by #3626 not be sufficient? CC @jimblandy, @gents83.

At the moment all the web resource deallocation is always in the hand of JS
At least all the specific_resource_drop() were implemented as empty function
As far as I can see today - with arcanization the resource will be inside an Arc till its needed on wgpu side, but if the latest one holding the resource is the user and he calls the drop on it - it'll be handled by JS

**Comment by cwfitzgerald** [comment]
> Except we implement another layer of reference counting mechanism in the wgpu layer, which is expensive and not a good solution.

<del>As things go it's probably not _that_ expensive, especially when only texture views and bind groups need to keep the resources alive.</del>

I keep getting fooled by this, it's invalid to submit a command buffer using a destroyed resource, so we'd need to keep the bind groups alive as well.

**Comment by grovesNL** [comment]
I think we should try to avoid diverging from the spec here. `drop` isn't strong enough intent to know that the resource isn't being used anymore. Instead people should call the explicit `destroy` function just like they would when using the web API directly.

**Comment by gents83** [comment]
Let me tackle and investigate it better while arcanizing 👍🏼

**Comment by JMS55** [comment]
@mikialex one thing to note at a high level, regardless of any issue in wgpu, is that you probably want to implement a texture cache (a simple hashmap will do) and keep textures around for 2-3 frames without any access before evicting them. This way you'll save on allocating textures in GPU memory every frame.

**Comment by mikialex** [comment]
@ErichDonGubler @gents83 https://github.com/gfx-rs/wgpu/pull/3626 This is a huge change but after some browsing, I suppose this PR will not help this issue. To make things work on the Web, If we do RAII-like resource management(destroy when dropped), we have to avoid destroying resources belonging to any recording but not submitted command buffer. What I mean by "implement another layer of reference counting mechanism in the wgpu layer" is that the encoding command buffer should keep strong references to all resources used to avoid the dropping when encoding. This should be done In wgpu layer, not in the wgpu-core. This approach is not feasible, because of the performance reason.

@grovesNL I suspect this is a spec issue. The current spec disallows RAII-like resource management except you take the recording command buffer into consideration.

@JMS55 We did have a texture pool for our use case, and if we use the texture pool this issue will not exposed so early. However, the real issue is you can not create large resources in high frequency and rely on the wgpu's drop to clean up in time. Only web target has this problem. This imposes a great limitation on the web target deployment because we have to check and add destroy calls manually. The other reason we do not use texture pool is that on some platforms we have correctness issues due to the driver bug. For example this: https://github.com/gfx-rs/wgpu/issues/2241 and another issue I have not submitted related to srgbness not working.

In our engine, all wgpu exposed resource is wrapped in Arc like container. So I could do the ref counting by myself. I'm working on another solution with the least performance overhead: If any command buffer instance exists, we globally defer any resource drop until all the command buffer instance has been submitted to the queue. Besides all these fixes above the wgpu, I'm not sure if the wgpu itself could or should implement a similar fix may be gated by a feature flag.

**Comment by cwfitzgerald** [comment]
> I suspect this is a spec issue. The current spec disallows RAII-like resource management except you take the recording command buffer into consideration.

I don't know what, if anything, the spec could do to improve this situation. What would you suggest?

**Comment by grovesNL** [comment]
I don't think there's anything actionable for this right now. We could reconsider some kind of optional refcounting at the WebGPU backend level sometime if it's a big problem. Maybe browsers could have some heuristics to run partial GC on held GPU resources after a command buffer submission or something.
